File: cropme/src/main/java/com/takusemba/cropme/ActionDetector.java
package com.takusemba.cropme;

import android.content.Context;
import android.support.v4.view.GestureDetectorCompat;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.ScaleGestureDetector;

/**
 * Detect all actions related to Image transition.
 **/
class ActionDetector {

    private ActionListener listener;
    private GestureDetectorCompat gestureDetectorCompat;
    private ScaleGestureDetector scaleGestureDetector;

    ActionDetector(Context context, ActionListener actionListener) {
        this.listener = actionListener;
        this.gestureDetectorCompat = new GestureDetectorCompat(context, simpleOnGestureListener);
        this.scaleGestureDetector = new ScaleGestureDetector(context, simpleScaleListener);
    }

    void detectAction(MotionEvent event) {
        if (gestureDetectorCompat == null) {
            throw new IllegalStateException("GestureDetectorCompat must not be null");
        }
        gestureDetectorCompat.onTouchEvent(event);
        scaleGestureDetector.onTouchEvent(event);
        switch (event.getAction()) {
            case MotionEvent.ACTION_UP:
                listener.onMoveEnded();
                break;
        }
    }

    private GestureDetector.SimpleOnGestureListener simpleOnGestureListener = new GestureDetector.SimpleOnGestureListener() {
        @Override
        public boolean onDown(MotionEvent e) {
            return true;
        }

        @Override
        public void onShowPress(MotionEvent e) {

        }

        @Override
        public boolean onSingleTapUp(MotionEvent e) {
            return true;
        }

        @Override
        public boolean onScroll(MotionEvent initialEvent, MotionEvent currentEvent, float dx, float dy) {
            listener.onMoved(-dx, -dy);
            return true;
        }

        @Override
        public void onLongPress(MotionEvent e) {

        }

        @Override
        public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
            listener.onFlinged(velocityX, velocityY);
            return true;
        }
    };

    private ScaleGestureDetector.SimpleOnScaleGestureListener simpleScaleListener
            = new ScaleGestureDetector.SimpleOnScaleGestureListener() {
        @Override
        public boolean onScaleBegin(ScaleGestureDetector detector) {
            return super.onScaleBegin(detector);
        }

        @Override
        public void onScaleEnd(ScaleGestureDetector detector) {
            super.onScaleEnd(detector);
            listener.onScaleEnded();
        }

        @Override
        public boolean onScale(ScaleGestureDetector detector) {
            listener.onScaled(detector.getScaleFactor());
            return true;
        }
    };
}

File: cropme/src/main/java/com/takusemba/cropme/ActionDetector.kt
package com.takusemba.cropme

import android.content.Context
import android.view.GestureDetector
import android.view.MotionEvent
import android.view.ScaleGestureDetector
import androidx.core.view.GestureDetectorCompat

/**
 * Detect all actions related to Image transition.
 */
internal class ActionDetector(
    context: Context,
    private val listener: ActionListener
) {

  private val gestureDetectorCompat: GestureDetectorCompat?
  private val scaleGestureDetector: ScaleGestureDetector

  private val simpleOnGestureListener = object : GestureDetector.SimpleOnGestureListener() {
    override fun onDown(e: MotionEvent): Boolean {
      return true
    }

    override fun onShowPress(e: MotionEvent) = Unit

    override fun onSingleTapUp(e: MotionEvent): Boolean {
      return true
    }

    override fun onScroll(
        initialEvent: MotionEvent, currentEvent: MotionEvent, dx: Float, dy: Float
    ): Boolean {
      listener.onMoved(-dx, -dy)
      return true
    }

    override fun onLongPress(e: MotionEvent) = Unit

    override fun onFling(
        e1: MotionEvent, e2: MotionEvent, velocityX: Float, velocityY: Float
    ): Boolean {
      listener.onFlinged(velocityX, velocityY)
      return true
    }
  }

  private val simpleScaleListener = object : ScaleGestureDetector.SimpleOnScaleGestureListener() {

    override fun onScaleBegin(detector: ScaleGestureDetector): Boolean {
      return super.onScaleBegin(detector)
    }

    override fun onScaleEnd(detector: ScaleGestureDetector) {
      super.onScaleEnd(detector)
      listener.onScaleEnded()
    }

    override fun onScale(detector: ScaleGestureDetector): Boolean {
      listener.onScaled(detector.scaleFactor)
      return true
    }
  }

  init {
    this.gestureDetectorCompat = GestureDetectorCompat(context, simpleOnGestureListener)
    this.scaleGestureDetector = ScaleGestureDetector(context, simpleScaleListener)
  }

  fun detectAction(event: MotionEvent) {
    if (gestureDetectorCompat == null) {
      throw IllegalStateException("GestureDetectorCompat must not be null")
    }
    gestureDetectorCompat.onTouchEvent(event)
    scaleGestureDetector.onTouchEvent(event)
    when (event.action) {
      MotionEvent.ACTION_UP -> listener.onMoveEnded()
    }
  }
}

